File: src/contracts/backend.py
from typing import Optional

from pydantic import BaseModel


class Backend(BaseModel):
    """
    Data model representing a backend service instance.
    """

    url: str
    port: Optional[int] = None
    health: bool = False
    in_flight_requests: float = 0.0
    avg_latency: float = 0.0

    def __eq__(self, other):
        """
        Check equality with another Backend instance based on URL and port.
        """
        if not isinstance(other, Backend):
            return False
        return self.url == other.url and self.port == other.port

    def __hash__(self):
        """
        Compute hash based on URL and port.
        """
        return hash((self.url, self.port))

    def __repr__(self):
        """
        Return a string representation of the Backend instance.
        """
        return (
            f"Backend(url={self.url}, port={self.port}, health={self.health}, "
            f"in_flight_requests={self.in_flight_requests}, avg_latency={self.avg_latency})"
        )

File: go/contracts/backend.go
package contracts

type Backend struct {
	URL              string  `json:"url"`
	Port             *int    `json:"port,omitempty"`
	Health           bool    `json:"health"`
	InFlightRequests float64 `json:"in_flight_requests"`
	AvgLatency       float64 `json:"avg_latency"`
}

func NewBackend(url string, port *int) *Backend {
	return &Backend{URL: url, Port: port, Health: false}
}

func (b *Backend) Equal(other *Backend) bool {
	if other == nil {
		return false
	}
	if b.URL != other.URL {
		return false
	}
	if (b.Port == nil) != (other.Port == nil) {
		return false
	}
	if b.Port != nil && other.Port != nil && *b.Port != *other.Port {
		return false
	}
	return true
}

// ModelDump returns a map similar to the Python model_dump() used elsewhere.
func (b *Backend) ModelDump() map[string]interface{} {
	var port interface{} = nil
	if b.Port != nil {
		port = *b.Port
	}
	return map[string]interface{}{
		"url":                b.URL,
		"port":               port,
		"health":             b.Health,
		"in_flight_requests": b.InFlightRequests,
		"avg_latency":        b.AvgLatency,
	}
}
